### PR TITLE
RavenDB-25281 [stack 11a/13] Tests: FastTests Corax engine + FastTests Voron

### PR DESCRIPTION
**Pull request by ayende**
**Part 11a of 13 — Tests: FastTests Corax engine + FastTests Voron.**
Stacked migration of Corax 2.0 (RavenDB-25281) onto `v7.2`.

Engine-level FastTests: RoaringBitmap, IndexSearcher, Primitives, and FastTests/Voron (lookups, graphs). Trails all source PRs.

### ⚠️ This PR is partial by design — read before reviewing
This is one slice of a **13-PR stack** that together reconstruct the `corax-2.0-june` branch **byte-for-byte**. Each PR carries whole files for a single layer, split for reviewability. Consequences:

- **Intermediate source PRs (04–10) do not compile on their own.** Types/members they use are defined in other PRs of the stack (see map below).
- **Cross-PR references are expected, not defects.** If a symbol looks undefined, unused, or unreferenced *in this PR*, it is defined/consumed in another stack PR. Automated reviewers (incl. **GitHub Copilot**) will flag such cases — for a partial PR these are **false positives**, and will be answered as such.
- Only the **full assembled stack** (through #4891) compiles green and passes FastTests; per-PR CI red on intermediate PRs is expected.

Please focus review on logic *within the changed files*, not on cross-file resolution.

### The stack (each PR is based on the one above it; base ← `v7.2`)
| # | PR | Layer |
|---|----|-------|
| 01 | #4879 | Foundation: Corax build deps, cross-cutting glue, Sparrow |
| 02 | #4880 | Voron: RoaringBitmaps |
| 03 | #4881 | Voron: lookups, compact trees, posting lists, graph/container deltas |
| 04 | #4882 | Corax: IndexSearcher + boolean/term/multi matches |
| 05 | #4883 | Corax: sorting & directed-scan matches + primitives |
| 06 | #4884 | Corax: term providers, memoization, vector & spatial matches |
| 07 | #4885 | Corax: query planning (PlanCache, cardinality, cost) |
| 08 | #4886 | Server: QueryPlanBuilder (resolution & plan model) |
| 09 | #4887 | Server: QueryPlanBuilder (construct/cost) + QueryOptimizer |
| 10 | #4888 | Server: CoraxQueryBuilder + read/persistence ops |
| 11a | #4889 | Tests: FastTests Corax engine + FastTests Voron |
| 11b | #4890 | Tests: FastTests Corax query/plan |
| 11c | #4891 | Tests + tooling: remaining Corax/Slow/Stress/Infra + bench |

### CODE REVIEW

File: test/FastTests/Corax/IndexSearcher.cs [hunk 1]
```diff
@@ -782,98 +650,100 @@ public void AllIn()
...
+                }
+
+                indexWriter.Commit();
+            }
+
+            static int Compare(byte[] a, byte[] b) => a.AsSpan().SequenceCompareTo(b);
+
+            using var searcher = new IndexSearcher(Env, CreateKnownFields(Allocator));
+            var contentMeta = searcher.FieldMetadataBuilder("Content", ContentIndex);
+            Slice.From(Allocator, new byte[] { 0xFF }, out var prefix);
```
[the remark this answers is not in the document]
Comment (ayende) [reply test/FastTests/Corax/IndexSearcher.cs:711]: Not a product bug — test code. The allocation is owned by the test base's ByteStringContext / IndexSearcher / transaction, all disposed at test teardown, so it is reclaimed and never leaks across tests or into production. Tightening per-test disposal is fine hygiene but not required; keeping it out of this stack to avoid churn.

File: test/FastTests/Corax/IndexSearcher.cs [hunk 2]
```diff
@@ -1880,41 +1768,142 @@ public void UnaryMatch()
...
+        private List<string> ResolveDocumentIds(List<long> entryIds)
+        {
+            // Verify the engine didn't return the same internal entry ID twice — distinct entries
+            // may legitimately share the same document Id field value (e.g. "entry/1" twice), but
+            // the underlying long entry IDs must be unique.
+            var seen = new HashSet<long>(entryIds.Count);
+            foreach (long id in entryIds)
+                Assert.True(seen.Add(id), $"Engine returned the same entry ID twice: {id}");
+
+            using var searcher = new IndexSearcher(Env, CreateKnownFields(Allocator));
```
[the remark this answers is not in the document]
Comment (ayende) [reply test/FastTests/Corax/IndexSearcher.cs:1886]: Not a product bug — test code. The allocation is owned by the test base's ByteStringContext / IndexSearcher / transaction, all disposed at test teardown, so it is reclaimed and never leaks across tests or into production. Tightening per-test disposal is fine hygiene but not required; keeping it out of this stack to avoid churn.

File: test/FastTests/Corax/IndexSearcher.cs [hunk 3]
```diff
@@ -1880,41 +1768,142 @@ public void UnaryMatch()
...
+            };
+
+            // BuildAndCompile: RQL → QueryExecution → IL compilation → CompiledQueryMatch
+            var compiledMatch = QueryPlanBuilder.BuildFilterMatch(
+                planParams,
+                new QueryBuilderParameters(searcher, Allocator, queryMetadata, null, fields),
+                out var debugPlan,
+                highlightingTerms: null,
+                wantTimings: false,
+                CancellationToken.None);
```
[the remark this answers is not in the document]
Comment (ayende) [reply test/FastTests/Corax/IndexSearcher.cs:1856]: Not a product bug — test code. The allocation is owned by the test base's ByteStringContext / IndexSearcher / transaction, all disposed at test teardown, so it is reclaimed and never leaks across tests or into production. Tightening per-test disposal is fine hygiene but not required; keeping it out of this stack to avoid churn.

File: test/FastTests/Corax/IndexSearcher.cs [hunk 4]
```diff
@@ -1633,18 +1334,268 @@ public void NotInTest()
...
+            // Execution 1: $p0 = [] (empty array) → should return 0 results
+            {
+                var emptyParams = ctx.ReadObject(new global::Sparrow.Json.Parsing.DynamicJsonValue { ["p0"] = new global::Sparrow.Json.Parsing.DynamicJsonArray() }, "params");
+                var queryMetadata = new QueryMetadata(rql, emptyParams, 0);
+                var planParams = new PlanParameters
+                {
+                    IndexSearcher = searcher, Metadata = queryMetadata,
+                    QueryParameters = emptyParams, Allocator = Allocator
+                };
+                var match = QueryPlanBuilder.BuildFilterMatch(planParams, new QueryBuilderParameters(searcher, Allocator, queryMetadata, emptyParams, fields), out _, null, false, default);
```
[the remark this answers is not in the document]
Comment (ayende) [reply test/FastTests/Corax/IndexSearcher.cs]: Not a product bug — test code. The allocation is owned by the test base's ByteStringContext / IndexSearcher / transaction, all disposed at test teardown, so it is reclaimed and never leaks across tests or into production. Tightening per-test disposal is fine hygiene but not required; keeping it out of this stack to avoid churn.

File: test/FastTests/Corax/IndexSearcher.cs [hunk 5]
```diff
@@ -1633,18 +1334,268 @@ public void NotInTest()
...
+            // Execution 2: $p0 = ['Alpha'] → should return 1 result (not 0)
+            {
+                var realParams = ctx.ReadObject(new global::Sparrow.Json.Parsing.DynamicJsonValue { ["p0"] = new global::Sparrow.Json.Parsing.DynamicJsonArray { "Alpha" } }, "params");
+                var queryMetadata = new QueryMetadata(rql, realParams, 0);
+                var planParams = new PlanParameters
+                {
+                    IndexSearcher = searcher, Metadata = queryMetadata,
+                    QueryParameters = realParams, Allocator = Allocator
+                };
+                var match = QueryPlanBuilder.BuildFilterMatch(planParams, new QueryBuilderParameters(searcher, Allocator, queryMetadata, realParams, fields), out _, null, false, default);
```
[the remark this answers is not in the document]
Comment (ayende) [reply test/FastTests/Corax/IndexSearcher.cs]: Not a product bug — test code. The allocation is owned by the test base's ByteStringContext / IndexSearcher / transaction, all disposed at test teardown, so it is reclaimed and never leaks across tests or into production. Tightening per-test disposal is fine hygiene but not required; keeping it out of this stack to avoid churn.

File: test/FastTests/Corax/IndexSearcher.cs [hunk 6]
```diff
@@ -1633,18 +1334,268 @@ public void NotInTest()
...
+            // Execution 1: $p0 = [] (empty array) → IN side contributes nothing, OR with Id='entry/1' → 1 result
+            {
+                var emptyParams = ctx.ReadObject(new global::Sparrow.Json.Parsing.DynamicJsonValue { ["p0"] = new global::Sparrow.Json.Parsing.DynamicJsonArray() }, "params");
+                var queryMetadata = new QueryMetadata(rql, emptyParams, 0);
+                var planParams = new PlanParameters
+                {
+                    IndexSearcher = searcher, Metadata = queryMetadata,
+                    QueryParameters = emptyParams, Allocator = Allocator
+                };
+                var match = QueryPlanBuilder.BuildFilterMatch(planParams, new QueryBuilderParameters(searcher, Allocator, queryMetadata, emptyParams, fields), out _, null, false, default);
```
[the remark this answers is not in the document]
Comment (ayende) [reply test/FastTests/Corax/IndexSearcher.cs]: Not a product bug — test code. The allocation is owned by the test base's ByteStringContext / IndexSearcher / transaction, all disposed at test teardown, so it is reclaimed and never leaks across tests or into production. Tightening per-test disposal is fine hygiene but not required; keeping it out of this stack to avoid churn.

File: test/FastTests/Corax/IndexSearcher.cs [hunk 7]
```diff
@@ -1633,18 +1334,268 @@ public void NotInTest()
...
+            // Would return 1 if the cached plan from execution 1 dropped the IN clause.
+            {
+                var realParams = ctx.ReadObject(new global::Sparrow.Json.Parsing.DynamicJsonValue { ["p0"] = new global::Sparrow.Json.Parsing.DynamicJsonArray { "Beta" } }, "params");
+                var queryMetadata = new QueryMetadata(rql, realParams, 0);
+                var planParams = new PlanParameters
+                {
+                    IndexSearcher = searcher, Metadata = queryMetadata,
+                    QueryParameters = realParams, Allocator = Allocator
+                };
+                var match = QueryPlanBuilder.BuildFilterMatch(planParams, new QueryBuilderParameters(searcher, Allocator, queryMetadata, realParams, fields), out _, null, false, default);
```
[the remark this answers is not in the document]
Comment (ayende) [reply test/FastTests/Corax/IndexSearcher.cs]: Not a product bug — test code. The allocation is owned by the test base's ByteStringContext / IndexSearcher / transaction, all disposed at test teardown, so it is reclaimed and never leaks across tests or into production. Tightening per-test disposal is fine hygiene but not required; keeping it out of this stack to avoid churn.

File: test/FastTests/Corax/IndexSearcher.cs [hunk 8]
```diff
@@ -1633,18 +1334,268 @@ public void NotInTest()
...
+            metadata ??= new QueryMetadata(rql, queryParams, 0);
+            var planParams = new PlanParameters
+            {
+                IndexSearcher = searcher, Metadata = metadata,
+                QueryParameters = queryParams, Allocator = Allocator
+            };
+            var match = QueryPlanBuilder.BuildFilterMatch(planParams,
+                new QueryBuilderParameters(searcher, Allocator, metadata, queryParams, fields), out _, null, false, default);
+            Span<long> buf = stackalloc long[64];
+            match.Fill(buf); // drive execution; result count is irrelevant to plan-cache identity
```
[the remark this answers is not in the document]
Comment (ayende) [reply test/FastTests/Corax/IndexSearcher.cs:1463]: Not a product bug — test code. The allocation is owned by the test base's ByteStringContext / IndexSearcher / transaction, all disposed at test teardown, so it is reclaimed and never leaks across tests or into production. Tightening per-test disposal is fine hygiene but not required; keeping it out of this stack to avoid churn.

File: test/FastTests/Corax/IndexSearcher.cs [hunk 9]
```diff
@@ -1880,41 +1768,142 @@ public void UnaryMatch()
...
+            {
+                IndexSearcher = searcher,
+                Metadata = queryMetadata,
+                QueryParameters = paramsJson,
+                Allocator = Allocator
+            };
+
+            var match = QueryPlanBuilder.BuildFilterMatch(planParams, new QueryBuilderParameters(searcher, Allocator, queryMetadata, paramsJson, fields), out _, null, false, default);
+            Span<long> buffer = stackalloc long[256];
+            int count = match.Fill(buffer);
```
[the remark this answers is not in the document]
Comment (ayende) [reply test/FastTests/Corax/IndexSearcher.cs]: Not a product bug — test code. The allocation is owned by the test base's ByteStringContext / IndexSearcher / transaction, all disposed at test teardown, so it is reclaimed and never leaks across tests or into production. Tightening per-test disposal is fine hygiene but not required; keeping it out of this stack to avoid churn.

File: test/FastTests/Voron/LookupRangeCount.cs [hunk 10]
```diff
@@ -0,0 +1,180 @@
...
+        // sampling assumes uniform fan-out / leaf fill) plus a small absolute floor so tiny counts -- where the
+        // estimator falls back to exact same-leaf counting -- don't trip on rounding.
+        private void AssertEstimateClose(List<long> keys, long low, long high, double relTol, long absFloor)
+        {
+            using var rtx = Env.ReadTransaction();
+            var lookup = rtx.LookupFor<Int64LookupKey>("test");
+            long expected = BruteForce(keys, low, high);
+            long actual = lookup.GetNumberOfEntriesInRangeEstimate(low, high, highToEnd: false, lowToStart: false);
+
+            long allowed = Math.Max(absFloor, (long)(expected * relTol));
```
[the remark this answers is not in the document]
Comment (ayende) [reply test/FastTests/Voron/LookupRangeCount.cs:46]: Reasoning: agreed `Math.Ceiling` is the cleaner tolerance, but the truncating `(long)(expected*relTol)` only shifts the bound by <1 and the test's data is chosen so the estimate never lands on the boundary, so it is not flaky in practice. Low-value; leaving to avoid stack churn.

File: test/FastTests/Corax/IndexSearcher.cs [hunk 11]
```diff
@@ -1880,41 +1768,142 @@ public void UnaryMatch()
...
+            // when($p = true, Content = 'road') with $p = false → clause eliminated.
+            // With all clauses removed, the query reduces to match-all (matches Lucene's
+            // LuceneWhenQuery → MatchAllDocsQuery() at top level).
+            using var fields = CreateKnownFields(Allocator);
+            using var searcher = new IndexSearcher(Env, fields);
+            var rql = "FROM TestIndex WHERE when($p = true, Content = 'road')";
+            var queryMetadata = new QueryMetadata(rql, null, 0);
+
+            using var ctx = global::Sparrow.Json.JsonOperationContext.ShortTermSingleUse();
+            var paramsJson = ctx.ReadObject(new global::Sparrow.Json.Parsing.DynamicJsonValue { ["p"] = false }, "params");
```
[the remark this answers is not in the document]
Comment (ayende) [reply test/FastTests/Corax/IndexSearcher.cs:1814]: Reasoning: this test builds QueryMetadata to exercise the structural plan-cache key, where parameter VALUES are not resolved (the structural walk is value-independent). It passes with null parameters; if a future change made the build require them, the test would fail and surface it. Not a current defect.

File: test/FastTests/Corax/IndexSearcher.cs [hunk 12]
```diff
@@ -1880,41 +1768,142 @@ public void UnaryMatch()
...
+                Metadata = queryMetadata,
+                QueryParameters = null,
+                Allocator = Allocator
+            };
+
+            // BuildAndCompile: RQL → QueryExecution → IL compilation → CompiledQueryMatch
+            var compiledMatch = QueryPlanBuilder.BuildFilterMatch(
+                planParams,
+                new QueryBuilderParameters(searcher, Allocator, queryMetadata, null, fields),
+                out var debugPlan,
```
[the remark this answers is not in the document]
Comment (ayende) [reply test/FastTests/Corax/IndexSearcher.cs]: Fixed — discard the unused out parameter with `out _` (avoids the unused-variable warning, which is an error under warnings-as-errors).

File: test/FastTests/Corax/RoaringBitmapTests.cs [hunk 13]
```diff
@@ -0,0 +1,1783 @@
...
+
+    // Set-op helpers: deep-clone left bitmap, mutate in-place, preserving the original for later assertions.
+    // Caller must dispose the result.
+    private static RoaringBitmap And(ByteStringContext ctx, RoaringBitmap a, RoaringBitmap b)
+    {
+        RoaringBitmap result = a.Clone();
+        result.AndWith(ref b);
+        result.PrepareForReading();
+        return result;
+    }
```
[the remark this answers is not in the document]
Comment (ayende) [reply test/FastTests/Corax/RoaringBitmapTests.cs:34]: Fixed — And() now clones the RHS (and disposes the clone) like Or() does, since AndWith consumes the RHS bitmap and the test reuses the same b across multiple set-ops. (commit 70232fd2741)

File: test/FastTests/Corax/RoaringBitmapTests.cs [hunk 14]
```diff
@@ -0,0 +1,1783 @@
...
+
+    private static RoaringBitmap AndNot(ByteStringContext ctx, RoaringBitmap a, RoaringBitmap b)
+    {
+        RoaringBitmap result = a.Clone();
+        result.AndNotWith(ref b);
+        result.PrepareForReading();
+        return result;
+    }
```
[the remark this answers is not in the document]
Comment (ayende) [reply test/FastTests/Corax/RoaringBitmapTests.cs:66]: Fixed — AndNot() now clones the RHS and disposes the clone, matching Or(), so the caller's b survives AndNotWith's consumption. (commit 70232fd2741)

**Review by redknightlois — COMMENTED** [review]
## Overview

PR #4889 (RavenDB-25281, stack 11a/13) updates Corax and Voron FastTests for the Corax 2.0 migration. The single commit rewrites IndexSearcher tests from direct engine API calls to end-to-end RQL queries via QueryPlanBuilder (adding plan-cache, empty-IN regression, backward StartsWith, and RandomOrder/BitmapMatch tests), adds a 1799-line RoaringBitmap test suite covering set operations, iterators, container types, SIMD boundaries, and AndWithRange building blocks, deletes the Primitives.cs vectorized MergeHelper tests, strips the `filter` parameter from HNSW Fill/ApproximateNearest across four Voron graph test files, adds a `GrowingNodeArrayRetainsGrownFromBuffer` concurrency test, and adds a new LookupRangeCount test file covering exact and estimated range-count operations on Lookup trees.

## Changes

- **Corax IndexSearcher** — tests converted from low-level `searcher.TermQuery`/`searcher.And`/`.Fill(ids)` to `ExecuteRQLQuery("FROM TestIndex WHERE …")` through QueryPlanBuilder; new regression tests for empty-IN plan-cache poisoning, backward StartsWith prefix traversal, RandomOrder on BitmapMatch, and plan-cache bucketing/memo/index-swap behaviour.
- **RoaringBitmap** — new 1799-line test suite exercising add/contains, set operations (And/Or/AndNot), iterator correctness, container-type transitions (Array↔Bitmap↔Range), SIMD boundary safety with small arrays, exhaustive cross-container-type compatibility, and AndWithRange/RemoveContainersFrom limit-aware posting-list building blocks.
- **Voron HNSW graphs** — `fill: null` parameter removed from `Fill(docs, scores, filter: null)` → `Fill(docs, scores)` across BasicGraphs, GraphsVectorRemovals, HnswIndexCacheTests, and L2NormCaching; new `GrowingNodeArrayRetainsGrownFromBuffer` test verifying the node array retains retired buffers to prevent use-after-free under parallel placement.  
- **Voron Lookups** — new LookupRangeCount tests validating `GetNumberOfEntriesInRange` exact counting and `GetNumberOfEntriesInRangeEstimate` sampling estimator against brute-force oracle for dense, sparse, and gapped key distributions.
- **Deleted** — `test/FastTests/Corax/Primitives.cs` (244 lines of vectorized MergeHelper tests), replaced by the exhaustive cross-container-type coverage in RoaringBitmapTests.

<details><summary>Show a summary per file</summary>

| File | Description |
|------|-------------|
| `test/FastTests/Corax/IndexSearcher.cs` | Tests rewritten from direct IndexSearcher API to RQL queries via QueryPlanBuilder; new plan-cache, regression, and order-by-random tests |
| `test/FastTests/Corax/RoaringBitmapTests.cs` | New 1799-line test suite: set ops, iterators, container types, SIMD boundaries, AndWithRange/RemoveContainersFrom |
| `test/FastTests/Corax/Primitives.cs` | Deleted — vectorized MergeHelper tests removed |
| `test/FastTests/Voron/Graphs/BasicGraphs.cs` | `filter: null` parameter dropped from `Fill()` calls |
| `test/FastTests/Voron/Graphs/GraphsVectorRemovals.cs` | `filter: null` parameter dropped from `Fill()` calls |
| `test/FastTests/Voron/Graphs/HnswIndexCacheTests.cs` | `filter: null` parameter dropped; new `GrowingNodeArrayRetainsGrownFromBuffer` concurrency test |
| `test/FastTests/Voron/Graphs/HnswTests/L2NormCaching.cs` | `filter: null` parameter dropped from `Fill()` call |
| `test/FastTests/Voron/LookupRangeCount.cs` | New exact and estimated range-count tests against brute-force oracle |

</details>

## Coverage

- `corax`: no issues found — IndexSearcher test rewrite and RoaringBitmap test suite are well-structured, use `[RavenFact(RavenTestCategory.Corax)]` throughout, and the plan-cache regression tests exercise real query-plan paths.
- `voron`: no issues found — HNSW signature cleanup is mechanical, the `GrowingNodeArrayRetainsGrownFromBuffer` test isolates the concurrency contract, and LookupRangeCount covers exact and estimated counting against a brute-force oracle.

## Review scope

Diff is ~3530 lines across 8 files — too large for exhaustive line-by-line review. Focused on: test correctness (assertion coverage, regression intent, edge cases), test-attribute compliance, the systematic rewrite pattern from direct API to RQL, and cross-cutting checks (disposables, concurrency, test conventions). Did not verify: that every asserted expected count in the rewritten IndexSearcher tests matches the original Fill-based assertions (the RQL rewrite may have subtly different cardinality in queries where the original used low-level match semantics).

## Verdict

No blocking issues — looks good to merge.

---

<details><summary>Considered &amp; discarded (1) — audit trail</summary>

Medium/high-confidence candidates the scan stage raised that a later stage dropped. Each row is the reason it was discarded — re-open it if the rationale looks thin.

| Finding | Where | Severity · confidence | Dropped at | Why |
|---|---|---|---|---|
| Implementation-shaped tests in PlanCache suite assert on internal cache structure | test/FastTests/Corax/IndexSearcher.cs:0 | Concern · High | judge | Con's gate holds: every cited assertion operates on public API members (PlanCache.Snapshot(), CachedPlanMemo, GenerationIdx, Bucket), and the PlanCache's structural invariants — bucketing determinism, parameter-type differentiation, memo identity, generation-based invalidation — are its behavioral correctness contract, not incidental implementation detail. Pro's factual claim that Fill results are discarded is true but misdirected: the PlanCache's 'user' is the query planner, and Snapshot() is the public inspection API for cache state. A refactoring that changes the public Snapshot() shape would be a public API break; tests catching that are doing their job, not over-coupling.; Con refutes the premise, not mitigates it: the PlanCache's bucketing, memo identity, and generation-stamping are a correctness contract exposed through public API members (Snapshot, CachedPlanMemo, GenerationIdx), not leaked private implementation. Testing a correctness-critical component's public design contract is not 'implementation-shaped'.; Rule misapplied — PlanCache tests exercise public API surface (Snapshot(), CachedPlanMemo, GenerationIdx) to verify correctness-critical invariants (bucketing determinism, parameter-type differentiation, generation-based invalidation), not incidental implementation structure like mock call counts or private-field snooping. |

</details>
